Reuse the authorization and template shared informers for GC

smarterclayton opened the pull request:
Allow them to be loaded via ForResource() as a stepping stone to
transforming their behavior.

For #8229

[test]

smarterclayton commented:
@liggitt I am confused why the preferred group versions are changing

liggitt commented:
I'm not sure either

smarterclayton commented:
[test]

smarterclayton commented:
It's because the generated informers require their API group to be registered.

smarterclayton commented:
Bolt db segmentation fault [test]

smarterclayton commented:
Fixed, review.

On Sat, Jun 3, 2017 at 10:00 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test SUCCESS (
> https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin/1920/)
> (Base Commit: 90592c5
> <https://github.com/openshift/origin/commit/90592c5815e371f648705c8dab384e36f21e8457>
> )
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/14391#issuecomment-306012574>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pwg0LhQxEwmHoYp69bFuf4vV8pVtks5sAg-ngaJpZM4NorhY>
> .
>

smarterclayton commented:
@enj also for review

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -580,6 +582,29 @@ func checkForOverrideConfig(ac configapi.AdmissionConfig) (*overrideapi.ClusterR
+	ForResource(resource schema.GroupVersionResource) (kinformers.GenericInformer, error)
+}
+
+type genericInformers struct {
```
This should be in a separate package.  I agree that combining informers is useful, but I suspect we'll need it more generally.

smarterclayton replied:
This was just the adapter interface for GC, if you want to start a brand new controller factory i want to do that in a separate PR where the rest of them get converted.

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -580,6 +582,29 @@ func checkForOverrideConfig(ac configapi.AdmissionConfig) (*overrideapi.ClusterR
+}
+
+type genericInformers struct {
+	kinformers.SharedInformerFactory
```
Why must this be anonymous?

smarterclayton replied:
this is passing through all the kubernetes informer methods

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -648,8 +673,11 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
+		ClientBuilder: saClientBuilder,
+		InformerFactory: genericInformers{
+			SharedInformerFactory: oc.Informers.KubernetesInformers(),
+			generic:               []GenericResourceInformer{oc.Informers},
```
Isn't this wiring in the hand coded informers?  I'm against including them in new places.

smarterclayton replied:
We need those to reuse existing informers

deads2k commented on pkg/controller/shared/shared_informer.go:
```diff
@@ -81,13 +87,26 @@ type sharedInformerFactory struct {
 	customListerWatchers  ListerWatcherOverrides
 	defaultResync         time.Duration
 
+	templateInformers      templateinformers.SharedInformerFactory
```
I don't like this. If we must have a new struct, we should be building out a new struct that doesn't carry the legacy of hand-coded informers with it.

smarterclayton replied:
What do you want the tsruct to look like?  The upstream shared struct?  This is the generated type

smarterclayton commented:
Refactored to be clearer, removed Register, and excluded a lot more resources that I honestly don't care about right now (mostly ones without reflectors).

smarterclayton commented:
[merge] based on feedback from previous

On Fri, Jun 9, 2017 at 3:06 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin/2066/)
> (Base Commit: 9279133
> <https://github.com/openshift/origin/commit/9279133f0a7b584aceae3c6275b8da91b1285655>
> )
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/14391#issuecomment-307474112>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6ZbehPkf25Bttuz7ZslNsrAie7jks5sCZelgaJpZM4NorhY>
> .
>
